File: ledgertimer/export.py
"""
export.py — CSV export logic for LedgerTimer.
"""

import csv
from datetime import datetime
from pathlib import Path
from typing import Sequence
import sqlite3


def _duration_decimal(start_iso: str, end_iso: str) -> str:
    """Return duration as decimal hours string, e.g. '1.75'."""
    fmt = "%Y-%m-%dT%H:%M:%S"
    start = datetime.strptime(start_iso, fmt)
    end = datetime.strptime(end_iso, fmt)
    hours = (end - start).total_seconds() / 3600
    return f"{hours:.2f}"


def export_to_csv(rows: Sequence[sqlite3.Row], filepath: str | Path) -> int:
    """
    Write rows to a CSV file.
    Returns the number of rows written.
    """
    fieldnames = ["Date", "Start Time", "End Time", "Duration", "Description", "Project"]

    with open(filepath, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        count = 0
        for row in rows:
            start_dt = datetime.strptime(row["start_time"], "%Y-%m-%dT%H:%M:%S")
            end_dt = datetime.strptime(row["end_time"], "%Y-%m-%dT%H:%M:%S")
            writer.writerow({
                "Date": start_dt.strftime("%Y-%m-%d"),
                "Start Time": start_dt.strftime("%H:%M"),
                "End Time": end_dt.strftime("%H:%M"),
                "Duration": _duration_decimal(row["start_time"], row["end_time"]),
                "Description": row["description"] or "",
                "Project": row["project"] or "",
            })
            count += 1

    return count

File: go/export.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"time"
)

// ExportToCSV writes log entries to w in CSV format.
// Returns the number of rows written.
func ExportToCSV(rows []LogEntry, w io.Writer) (int, error) {
	cw := csv.NewWriter(w)
	if err := cw.Write([]string{"Date", "Start Time", "End Time", "Duration", "Description", "Project"}); err != nil {
		return 0, err
	}
	count := 0
	for _, row := range rows {
		start, err := time.Parse(dtFmt, row.StartTime)
		if err != nil {
			continue
		}
		end, err := time.Parse(dtFmt, row.EndTime)
		if err != nil {
			continue
		}
		hours := end.Sub(start).Hours()
		if err := cw.Write([]string{
			start.Format("2006-01-02"),
			start.Format("15:04"),
			end.Format("15:04"),
			fmt.Sprintf("%.2f", hours),
			row.Description,
			row.Project,
		}); err != nil {
			return count, err
		}
		count++
	}
	cw.Flush()
	return count, cw.Error()
}
